### PR TITLE
Allow RNN to work with categorical features

### LINKED ISSUE #323: Allow RNN to work with categorical features
### 🚀 Feature Request

Add possibility to use categorical feature while training RNN model.

### Proposal

Required changes:

- Add new parameter `embedding_sizes` of type `Dict`, where
  - Keys are names of categorical features and 
  - Features are tuple of (number of unique values, feature embedding_size)
- Implement common class `MultiEmbedding` like this:
```
Class MultiEmbedding(nn.Module):
    def __init__(
        self,
        embedding_sizes: Dict[str, Tuple[int, int]],
    ):
        """
        embedding_sizes: dictionary mapping feature name to tuple of number of categorical classes and embedding size
        """
    ...
     
    def forward(x: Dict[str, torch.Tensor]) -> Dict[str, torch.Tensor]:
        """
        x: dictionary mapping feature name to feature values
        """
```
- Change `make_samples` method to handle new categorical values (like in TFT Native) and add categorical features to batch
- Change `forward` and `step` methods to get embeddings of categorical features 
 

### Test cases

- RNN works correctly with categorical features
  - Features that have unseen values in future

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by egoriyaa**
## Before submitting (must do checklist)

- [ ] Did you read the [contribution guide](https://github.com/etna-team/etna/blob/master/CONTRIBUTING.md)?
- [ ] Did you update the docs? We use Numpy format for all the methods and classes.
- [ ] Did you write any new necessary tests?
- [ ] Did you update the [CHANGELOG](https://github.com/etna-team/etna/blob/master/CHANGELOG.md)?

## Closing issues
closes #323

### CODE REVIEW

File: etna/models/nn/rnn.py [hunk 1]
```diff
@@ -244,6 +298,8 @@ def __init__(
             number of layers
         hidden_size:
             size of the hidden state
+        embedding_sizes:
```
Comment (d-a-bunin) [inline etna/models/nn/rnn.py:301]: I don't see any comments about handling of embeddings:
- Working with label-encoded values
- Handling unknown values

File: etna/models/nn/utils.py [hunk 2]
```diff
@@ -310,3 +312,34 @@ def _make_target_prediction(self, ts: TSDataset, horizon: int) -> Tuple[TSDatase
...
+        Parameters
+        ----------
+        embedding_sizes:
+            dictionary mapping feature name to tuple of number of categorical classes and embedding size
+        """
+        super().__init__()
+
+        self.embedding_sizes = embedding_sizes
+        self.embedding = nn.ModuleDict(
+            {feature: nn.Embedding(n + 1, dim) for feature, (n, dim) in self.embedding_sizes.items()}
```
Comment (d-a-bunin) [inline etna/models/nn/utils.py:333]: Why n+1?

File: tests/test_models/test_nn/test_rnn.py [hunk 3]
```diff
@@ -45,10 +48,37 @@ def test_rnn_model_run_weekly_overfit_with_scaler(ts_dataset_weekly_function_wit
     assert mae(ts_test, future) < 0.06
 
 
+@pytest.mark.parametrize(
+    "embedding_sizes,features_to_encode",
+    [({}, []), ({"categ_regr_label": (2, 5), "categ_regr_new_label": (1, 5)}, ["categ_regr", "categ_regr_new"])],
+)
+def test_rnn_backtest(ts_different_regressors, embedding_sizes, features_to_encode):
```
Comment (d-a-bunin) [inline tests/test_models/test_nn/test_rnn.py:55]: Are we sure this test is about backtest and not about handling categoricals?

File: tests/test_models/test_nn/test_rnn.py [hunk 4]
```diff
@@ -45,10 +48,37 @@ def test_rnn_model_run_weekly_overfit_with_scaler(ts_dataset_weekly_function_wit
...
+        + [FilterFeaturesTransform(exclude=["reals_exog", "categ_exog"])],
+        horizon=1,
+    )
+    pipeline.backtest(ts_different_regressors, metrics=[MAE()], n_folds=2)
+
+
+@pytest.mark.parametrize("categ_columns", [[], ["regressor_int_cat"]])
 @pytest.mark.parametrize("df_name", ["example_make_samples_df", "example_make_samples_df_int_timestamp"])
-def test_rnn_make_samples(df_name, request):
+def test_rnn_make_samples(df_name, categ_columns, request):
```
Comment (d-a-bunin) [inline tests/test_models/test_nn/test_rnn.py:79]: I don't like this short name: `categ_columns`. Let's use `cat_columns` or `category_columns`.

File: etna/models/nn/utils.py [hunk 5]
```diff
@@ -310,3 +312,34 @@ def _make_target_prediction(self, ts: TSDataset, horizon: int) -> Tuple[TSDatase
         ts.df = ts.df.iloc[-horizon:]
         ts.loc[:, pd.IndexSlice[:, "target"]] = predicts.T[:horizon]
         return ts, prediction_dataloader
+
+
+class MultiEmbedding(nn.Module):
```
Comment (d-a-bunin) [inline etna/models/nn/utils.py:317]: Shouldn't we write tests about this class?

**Review by d-a-bunin — CHANGES_REQUESTED** [review]

File: examples/202-NN_examples.ipynb [hunk 6]
```diff
@@ -55,7 +55,7 @@
   },
```
Comment (d-a-bunin) [inline examples/202-NN_examples.ipynb:55]: <pre class="ql-syntax" spellcheck="false">Line #8.    transform_date = DateFlagsTransform(day_number_in_week=True, day_number_in_month=False, out_column="dateflag")
</pre><p>Technically, there will be another feature with true value because of the default values.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/etna-team/etna/pull/334/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='82'/>

File: etna/models/nn/rnn.py [hunk 7]
```diff
@@ -142,11 +176,18 @@ def make_samples(self, df: pd.DataFrame, encoder_length: int, decoder_length: in
             .pipe(lambda x: x[["target_shifted"] + [i for i in x.columns if i != "target_shifted"]])
```
[the remark this answers is not in the document]
Comment (egoriyaa) [reply etna/models/nn/rnn.py:176]: There are two ways:
1. Throw exception
2. Throw warning and delete this feature from `embedding_sizes`

File: etna/models/nn/rnn.py [hunk 8]
```diff
@@ -142,11 +176,18 @@ def make_samples(self, df: pd.DataFrame, encoder_length: int, decoder_length: in
...
             .values
         )
+        for feature in self.embedding_sizes:
+            df[feature] = df[feature].astype(float).fillna(self.embedding_sizes[feature][0])
```
[the remark this answers is not in the document]
Comment (egoriyaa) [reply etna/models/nn/rnn.py:180]: It is ok

File: etna/models/nn/utils.py (hunk 2 above)
Comment (egoriyaa) [reply etna/models/nn/utils.py:333]: We have discussed that we need one more embedding for unknown category

Comment (d-a-bunin) [reply etna/models/nn/utils.py:333]: We should clarify it using the comment and may be write about this in documentation of the class.

File: tests/test_models/test_nn/test_rnn.py (hunk 3 above)
Comment (egoriyaa) [reply tests/test_models/test_nn/test_rnn.py:55]: I was surprised that we don't have backtest test, so I added it to check whether it works in all main setups

File: examples/202-NN_examples.ipynb (hunk 6 above)
Comment (egoriyaa) [reply examples/202-NN_examples.ipynb:55]: It is not so important because we specify exact category columns

File: etna/models/nn/utils.py (hunk 5 above)
Comment (egoriyaa) [reply etna/models/nn/utils.py:317]: I will add

File: etna/models/nn/rnn.py (hunk 1 above)
Comment (egoriyaa) [reply etna/models/nn/rnn.py:305]: Added

File: etna/models/nn/rnn.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (egoriyaa) [reply etna/models/nn/rnn.py:181]: Added

File: etna/models/nn/utils.py (hunk 2 above)
Comment (egoriyaa) [reply etna/models/nn/utils.py:333]: I think we don't need to write about deep implementation nuances in documentation of the class.
I will write comment  above this line

File: etna/models/nn/rnn.py [hunk 9]
```diff
@@ -206,6 +263,10 @@ def configure_optimizers(self) -> "torch.optim.Optimizer":
 class RNNModel(DeepBaseModel):
     """RNN based model on LSTM cell.
 
+    Model needs label encoded inputs for categorical features, for that purposes use :py:class:`~etna.transforms.LabelEncoderTransform`.
+    Feature values that were not seen during `fit` should be set to NaN for expected behaviour with `strategy="none"`.
+
+    If there are numeric columns that are passed to `embedding_sizes` parameter, they will be considered only as categorical features.
```
Comment (d-a-bunin) [inline etna/models/nn/rnn.py:269]: You should add an empty line after this one.

Comment (d-a-bunin) [reply etna/models/nn/rnn.py:269]: `embedding_sizes` should be in double ``.

File: etna/models/nn/rnn.py [hunk 10]
```diff
@@ -206,6 +263,10 @@ def configure_optimizers(self) -> "torch.optim.Optimizer":
 class RNNModel(DeepBaseModel):
     """RNN based model on LSTM cell.
 
+    Model needs label encoded inputs for categorical features, for that purposes use :py:class:`~etna.transforms.LabelEncoderTransform`.
+    Feature values that were not seen during `fit` should be set to NaN for expected behaviour with `strategy="none"`.
```
Comment (d-a-bunin) [inline etna/models/nn/rnn.py:267]: > for expected behaviour with `strategy="none"`.

Sounds strange, may be better to write:
```
Feature values that weren't seen during `fit` should be set to NaN, to get this behaviour use encoder with `strategy="none"`.
```
